codalab/worker/docker_image_manager.py

```diff
@@ -171,7 +171,18 @@ def image_availability_state(image_spec, success_message, failure_message):
                     tag_label, timestamp = tag.split(":")
                     # remove any other timestamp but not the current one
                     if tag_label == self.CACHE_TAG and timestamp != new_timestamp:
-                        self._docker.images.remove(tag)
+                        try:
+                            self._docker.images.remove(tag)
+                        except docker.errors.ImageNotFound as err:
+                            # It's possible that we get a 404 not found error here when removing the image,
```

Comment:
Just curious about this error, coz from docker-py documentation, it doesn't document this particular error in the [remove](https://docker-py.readthedocs.io/en/stable/images.html#docker.models.images.ImageCollection.remove) option, although this can totally be their mistake. I just wonder will there be any other error which could lead to a similar issue as well?

Reply:
Here's where i got it from:

```
(cl_dev) nfliu at mazama in ~/git/codalab-worksheets on master
$ pip list | grep docker
docker              3.7.0
docker-pycreds      0.4.0
(cl_dev) nfliu at mazama in ~/git/codalab-worksheets on master
$ ipython
Python 3.6.10 |Anaconda, Inc.| (default, Jan  7 2020, 15:01:53)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import docker

In [2]: docker = docker.from_env()

In [3]: docker.images.remove("animagethatdoesntexist")
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/api/client.py in _raise_for_status(self, response)
    255         try:
--> 256             response.raise_for_status()
    257         except requests.exceptions.HTTPError as e:

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/requests/models.py in raise_for_status(self)
    940         if http_error_msg:
--> 941             raise HTTPError(http_error_msg, response=self)
    942

HTTPError: 404 Client Error: Not Found for url: http+docker://localhost/v1.35/images/animagethatdoesntexist?force=False&noprune=False

During handling of the above exception, another exception occurred:

ImageNotFound                             Traceback (most recent call last)
<ipython-input-3-9ab3846e26db> in <module>
----> 1 docker.images.remove("animagethatdoesntexist")

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/models/images.py in remove(self, *args, **kwargs)
    457
    458     def remove(self, *args, **kwargs):
--> 459         self.client.api.remove_image(*args, **kwargs)
    460     remove.__doc__ = APIClient.remove_image.__doc__
    461

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/utils/decorators.py in wrapped(self, resource_id, *args, **kwargs)
     17                     'Resource ID was not provided'
     18                 )
---> 19             return f(self, resource_id, *args, **kwargs)
     20         return wrapped
     21     return decorator

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/api/image.py in remove_image(self, image, force, noprune)
    479         params = {'force': force, 'noprune': noprune}
    480         res = self._delete(self._url("/images/{0}", image), params=params)
--> 481         return self._result(res, True)
    482
    483     def search(self, term):

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/api/client.py in _result(self, response, json, binary)
    260     def _result(self, response, json=False, binary=False):
    261         assert not (json and binary)
--> 262         self._raise_for_status(response)
    263
    264         if json:

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/api/client.py in _raise_for_status(self, response)
    256             response.raise_for_status()
    257         except requests.exceptions.HTTPError as e:
--> 258             raise create_api_error_from_http_exception(e)
    259
    260     def _result(self, response, json=False, binary=False):

~/miniconda3/envs/cl_dev/lib/python3.6/site-packages/docker/errors.py in create_api_error_from_http_exception(e)
     29         else:
     30             cls = NotFound
---> 31     raise cls(e, response=response, explanation=explanation)
     32
     33

ImageNotFound: 404 Client Error: Not Found ("No such image: animagethatdoesntexist:latest")
```

It's possible that other exceptions could lead to this...but we can add those exceptions later if need be?